src/librustc_typeck/check/mod.rs

```diff
@@ -1104,6 +1110,20 @@ fn typeck_with_fallback<'tcx>(
         // refer to opaque types.
         fcx.select_obligations_where_possible(fallback_has_occurred, |_| {});
 
+        // We run through the list of `unsolved_variables` gathered earlier and check if there
+        // are any that are marked `Diverging` at this point. if so, this would imply, that they
+        // were assigned Divergent type because of fallback.
+        //
+        // Any type in `unsolved_variables` that is now `!`, is `!` as a result of fallback.
+        let mut from_diverging_fallback = unsolved_variables;
+        let diverging_default = fcx.tcx.mk_diverging_default();
+        from_diverging_fallback
+            .retain(
+                |ty|
+                    fcx.infcx.resolve_vars_if_possible(ty) == diverging_default &&
+                        !fcx.dead_nodes.borrow().contains(&id)
```

Comment:
what is `id` here? I don't think this is quite right-- I think we're going to want to test for whether something is a dead node later on, during writeback

Reply:
Makes sense.